harness/determined/keras/callbacks.py

```diff
@@ -0,0 +1,565 @@
+        self._corrected_test_end(logs)
+
+    # Must be triggered externally.
+    def _train_workload_end(self, metrics: Dict) -> None:
+        self.state.latest_training_metrics = metrics
+        for cb in self.callbacks:
+            if isinstance(cb, Callback):
+                cb.on_train_workload_end(self.state.total_batches, metrics)
+        self.want_train_workload_begin = True
+
+    # Must be triggered externally.
+    def _get_state(self) -> Dict:
+        cb_state = [
+            cb.get_state() if isinstance(cb, Callback) else None for cb in self.all_callbacks
+        ]
+        return {"self": vars(self.state), "callbacks": cb_state}
+
+    # Must be triggered externally.
+    def _checkpoint_end(self, checkpoint_dir: str) -> None:
+        for cb in self.callbacks:
+            if isinstance(cb, Callback):
+                cb.on_checkpoint_end(checkpoint_dir)
+
+    # Must be triggered externally.
+    def _load_state(self, state: Dict) -> None:
+        self.state = self.SaveableState(**state["self"])
+        for cb, cb_state in zip(self.all_callbacks, state["callbacks"]):
+            if isinstance(cb, Callback):
+                cb.load_state(cb_state)
+
+
+class DeterminedProgress(Callback):
+    """A Determined-friendly replacement for the usual verbose=1 behavior."""
+
+    _chief_worker_only = True
+    _supports_tf_logs = True
+
+    def __init__(self) -> None:
+        self.batches_requested = None  # type: Optional[int]
+        self.batches_trained = 0
+        self.percent_reported = -1
+
+    def _report(self, logs: Optional[Dict]) -> None:
+        # Only report progress if we have batches_requested information.
+        if self.batches_requested is None:
+            return
+
+        # Don't report more often than 10% increments.
+        percent_10 = int((self.batches_trained / self.batches_requested) * 10) * 10
+        if percent_10 <= self.percent_reported:
+            return
+
+        # When you do report, report at 1% increments.
+        percent = int((self.batches_trained / self.batches_requested) * 100)
+
+        report = (
+            f"total batches trained: {self.total_batches_trained}, "
+            f"workload {percent}% complete ({self.batches_trained}/{self.batches_requested})"
+        )
+        if logs is not None:
+            metrics = {k: v for k, v in logs.items() if k not in ("batch", "size")}
+            report += f": {metrics}"
+
+        print(report)
+
+    def on_train_workload_begin(
+        self, total_batches_trained: int, batches_requested: Optional[int], logs: Dict
+    ) -> None:
+        self.batches_requested = batches_requested
+        self.batches_trained = 0
+        self.total_batches_trained = total_batches_trained
+        self._report(None)
+
+    def on_train_batch_end(self, batch: int, logs: Dict) -> None:
+        self.batches_trained += 1
+        self.total_batches_trained += 1
+        self._report(logs)
+
+
+class DeterminedHistory(Callback):
```

Comment:
non-blocking: should this be `_DeterminedHistory` since users will never be using themselves ?

Reply:
done.